lib/translator/inquiry.cc

```diff
@@ -187,6 +187,109 @@ void TranslateUnitSerialNumberVpd(
   memcpy(&buffer[sizeof(result)], &product_serial_number, result.page_length);
 }
 
+void TranslateLogicalBlockProvisioningVpd(
+    const nvme_defs::IdentifyControllerData &identify_controller_data,
+    const nvme_defs::IdentifyNamespace &identify_namespace_data,
+    absl::Span<uint8_t> buffer) {
+  bool ad = identify_controller_data.oncs.dsm;
+
+  scsi_defs::LogicalBlockProvisioningVpd result = {
+      .page_code = scsi_defs::PageCode::kBlockLimitsVpd,
+      .page_length = 0x04,
+
+      /*
+      THRESHOLD_EXPONENT
+      Shall be set to 00h to indicate that there are no thin
+      provisioning thresholds
+      This would require modification if thin-provisioning is
+      supported
+
+      TODO:
+      Right now we assume this field is 0.
+      If thin provisioning is enabled, we can calculate a valid threshold
+      exponnet from the READ CAPACITY (16) parameter data.
+      */
+
+      // .threshold_exponent= identify_namespace_data.nsfeat.thin_prov ?
+      // read_capacity : 0,
+
+      /*
+      LBPRZ
+      Shall be set to 1 if Dataset Management command –
+      Deallocate (AD) attribute is supported and the device
+      returns all zeros for reads of deallocated LBAs, otherwise
+      set to 0.
+
+      TODO: the AD attribute itself may not guarantee the device
+      returns all zeros for reads of deallocated LBAs.
+      May have to double check this one if there are any errors.
+      */
+      .lbprz = ad,
+
+      /*
+      ANC_SUP
+      Shall be set to 0, to indicate that setting the ANCHOR bit
+      in UNMAP is not supported if the namespace is not
+      resource or thin provisioned.
+
+      Shall be set to 1, to indicate that setting the ANCHOR bit
+      in UNMAP is supported if the namespace is resource provisioned.
+
+      May be set to 1 if the namespace is thin provisioned.
+
+      TODO: Lack of support from thin provisioning by itself may not
+      guarantee lack of resource provisioning. There may be an explicit flag
+      about resource provisioning.
+      */
+
+      .anc_sup = identify_namespace_data.nsfeat.thin_prov,
+
+  };
+
+  /*
+  PROVISIONING TYPE
+  Shall be set to 0 (Full) if Dataset Management command
+  – Deallocate (AD) attribute is not supported and Identify
+  Namespace Data = NSFEAT bit 0 is reported “0”
+  indicating that the namespace is not thin-provisioned.
+
+  Shall be set to 1(Resource) if Dataset Management
+  command – Deallocate (AD) attribute is supported and
+  Identify Namespace Data – NSFEAT bit 0 is reported “0”,,
+  indicating that the namespace is resource-provisioned.
+
+  Shall be set to 2(Thin) if Dataset Management command
+  – Deallocate (AD) attribute is supported and Identify
+  Namespace Data – NSFEAT bit 0 is reported “1”,
+  indicating that the namespace is thin-provisioned.
+   */
+  if (!ad && !identify_namespace_data.nsfeat.thin_prov)
```

Comment:
I have a strong personal preference against C/C++ code using if/else with no braces, but I don't know that it officially violates the style rules you're using for this project.

Reply:
I would either add braces, or remove the blank lines between clauses. Blank lines in code tend to create a mental disconnect between the stuff above and below, but in if/else clauses it should be clear which clauses are linked to each other.